ELF loader: Refactor and fix multiple issues.

mato opened the pull request:
@cfcs found multiple issues in our ELF loader (#382, #383). This code is quite old and has been in need of review for some time, so here is a first cut of refactoring along with fixes for the reported issues.

See individual commits for details.

Fixes #382, fixes #383, fixes #384.

cfcs commented on tenders/common/elf.c:
```diff
@@ -73,120 +74,155 @@ static ssize_t pread_in_full(int fd, void *buf, size_t count, off_t offset)
+         * Load the segment into host memory at host_vaddr and ensure any BSS
+         * (p_memsz - p_filesz) is initialised to zero.
+         */
+        uint8_t *host_vaddr = mem + p_vaddr;
```
1) if `mem_size` is large (in turn permitting a large `p_vaddr`), and `mem` (allocated by the caller) is near the end of the address space, doesn't this risk overflowing?

2) Is it intentional that the start address is not adjusted for alignment, and that the only influence that `p_align` has is on `p_vaddr_end` (and in turn for the range for the `mprotect()` call)?

mato replied:
> If mem_size is large (in turn permitting a large p_vaddr), and mem (allocated by the caller) is near the end of the address space, doesn't this risk overflowing?

I don't see how the caller could allocate a valid memory range (mem, memsize) where (mem + memsize) overlaps the address space. In any case, that would seem to be the caller's problem, not the ELF loader's?

> Is it intentional that the start address is not adjusted for alignment, and that the only influence that p_align has is on p_vaddr_end (and in turn for the range for the mprotect() call)?

No idea, I should re-read that part of the ELF spec.

cfcs replied:
1) Yes, that is the caller's problem. But `mem_size` comes from the user, and `mem` is allocated by `malloc`, so I wonder if we should have something to handle this in place somewhere? (I don't know if we do). It seems like a failure mode that would be worth either checking for here, or documenting in `elf.h`?

mato replied:
(1) Done in abeca34 using `assert()`s since this is a caller error.

cfcs commented on tenders/common/elf.c:
```diff
@@ -73,120 +74,155 @@ static ssize_t pread_in_full(int fd, void *buf, size_t count, off_t offset)
-        else {
-            _end = result;
+            p_vaddr_end = p_vaddr_end & -p_align;
         }
```
I'm not 100% sure I'm following the bit hacking here, but it seems to me that if `p_align` is a multiple of 2 then: 
```c
Elf64_Addr  p_vaddr_end = ((p_vaddr + p_memsz) + (p_align -1)) & (-p_align);
```
It has been verified previously that `p_vaddr + p_filesz < mem_size` and that `p_vaddr + p_memsz < mem_size`, but I'm wondering if a large `p_align` can lead us change the memory protection flags of a range that goes beyond `mem_size` ?

mato replied:
Good question. Also not obvious if we shouldn't be doing the `memset()` using the size adjusted for alignment. WIll check.

cfcs commented on tenders/common/elf.c:
```diff
@@ -73,120 +74,155 @@ static ssize_t pread_in_full(int fd, void *buf, size_t count, off_t offset)
-                || (result >= mem_size))
+        if ((p_vaddr >= mem_size)
+                || add_overflow(p_vaddr, p_filesz, p_vaddr_end)
+                || (p_vaddr_end >= mem_size))
```
Hmm, re-reading this, shouldn't it be `> mem_size` instead of `>= mem_size`?
Consider the case where someone has a mapping of 2MB - or a single byte, for simplicity:
```c
p_vaddr = 0;
p_filesz = p_memsz = 1;
mem_size = 1;
/* it follows that */
!( p_vaddr >= mem_size ); /* good, within range */
p_vaddr_end = 0 + 1; /* good, no overflow */
p_vaddr_end >= 1; /* rejected since p_vaddr_end == 1 */
```
Now p_vaddr_end is

mato replied:
Done in abeca34.

mato commented:
@cfcs Ok, I've done another pass in abeca34. This makes the ELF loader even more strict -- see the commit message for details. I've also implemented adjusting the segment start address for alignment.

I also deliberately disallow overlapping segments in the ELF loader. As a side effect of this, I found that clang (FreeBSD) was generating such an executable for `test_tls` due to missing TLS sections in our linker script, so I fixed that -- I think -- at least enough to make the test pass. I'm still not convinced that we are actually doing enough/the right thing regarding TLS but don't want to get too distracted by that particular rabbit hole right now.

Please review.

mato commented:
Note that the OpenBSD test is actually successful, but is showing as pending due to random surf- build/node.js brokenness.

cfcs commented on tenders/common/elf.c:
```diff
@@ -73,120 +75,237 @@ static ssize_t pread_in_full(int fd, void *buf, size_t count, off_t offset)
+            goto out_invalid;
+        if (p_vaddr_end & (EM_PAGE_SIZE - 1))
+            goto out_invalid;
+        host_vaddr = mem + p_vaddr_start;
```
I find the reuse of this variable (`host_vaddr`) a bit confusing.
In the blocks above, `host_vaddr = mem + p_vaddr` is being used to load `p_filesz` bytes (unaligned).
Then additional stuff after that is `memset` up until length `mem + p_memsz`.

In this line here it is reassigned as `host_vaddr = mem + p_vaddr_start` (aligned, rounded down), after which the mapping between `mem + v_addr_start` and `mem + align_up(p_memsz, p_align)` is `mprotect()'ed` using the desired flags.
What is the intended contents of `mem + align_down(p_vaddr, p_align)` through `mem + p_vaddr` at this point? Should that also be cleared, or is the caller expected to have cleared `mem ... mem+mem_size`?

mato replied:
> I find the reuse of this variable (host_vaddr) a bit confusing.

Fair enough. How about I introduce a `host_vaddr_start` for the second block?

> What is the intended contents of mem + align_down(p_vaddr, p_align) through mem + p_vaddr at this point?

The [ELF specification](http://refspecs.linuxfoundation.org/elf/elfspec.pdf) discusses this on page 2-8:

> ‘‘Impurities’’ in the other three pages are not logically part of the process image; whether the system expunges them is unspecified.

I.e. we don't care. In practice, given that the host memory comes from `mmap()` it will on all modern systems be initialised to zero.
